src/Fixer/PhpUnit/PhpUnitDataProviderReturnTypeFixer.php

```diff
@@ -0,0 +1,119 @@
+use PhpCsFixer\Tokenizer\CT;
+use PhpCsFixer\Tokenizer\Token;
+use PhpCsFixer\Tokenizer\Tokens;
+
+/**
+ * @author Kuba Werłos <werlos@gmail.com>
+ */
+final class PhpUnitDataProviderReturnTypeFixer extends AbstractPhpUnitFixer
+{
+    public function getDefinition(): FixerDefinitionInterface
+    {
+        return new FixerDefinition(
+            'The type returned by data provider must be `iterable`.',
+            [
+                new CodeSample(
+                    '<?php
+class FooTest extends TestCase {
+    /**
+     * @dataProvider provideSomethingCases
+     */
+    public function testSomething($expected, $actual) {}
+    public function provideSomethingCases(): array {}
+}
+',
+                ),
+                new CodeSample(
+                    '<?php
+class FooTest extends TestCase {
+    /**
+     * @dataProvider provideSomethingCases
+     */
+    public function testSomething($expected, $actual) {}
+    public function provideSomethingCases() {}
+}
+',
+                ),
+            ],
+            null,
+            'Risky when relying on signature of the data provider.',
+        );
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * Must run before ReturnTypeDeclarationFixer.
+     * Must run after CleanNamespaceFixer.
+     */
+    public function getPriority(): int
+    {
+        return 0;
+    }
+
+    public function isRisky(): bool
+    {
+        return true;
+    }
+
+    protected function applyPhpUnitClassFix(Tokens $tokens, int $startIndex, int $endIndex): void
+    {
+        $dataProviderAnalyzer = new DataProviderAnalyzer();
+        $functionsAnalyzer = new FunctionsAnalyzer();
+
+        foreach (array_reverse($dataProviderAnalyzer->getDataProviders($tokens, $startIndex, $endIndex)) as $dataProviderAnalysis) {
+            $typeAnalysis = $functionsAnalyzer->getFunctionReturnType($tokens, $dataProviderAnalysis->getNameIndex());
+
+            if (null === $typeAnalysis) {
+                $argumentsStart = $tokens->getNextTokenOfKind($dataProviderAnalysis->getNameIndex(), ['(']);
+                $argumentsEnd = $tokens->findBlockEnd(Tokens::BLOCK_TYPE_PARENTHESIS_BRACE, $argumentsStart);
+
+                $tokens->insertAt(
+                    $argumentsEnd + 1,
+                    [
+                        new Token([CT::T_TYPE_COLON, ':']),
+                        new Token([T_WHITESPACE, ' ']),
+                        new Token([T_STRING, 'iterable']),
+                    ],
+                );
+
+                continue;
+            }
+
+            if ('iterable' === $typeAnalysis->getName()) {
+                continue;
+            }
+
+            $typeStartIndex = $tokens->getNextMeaningfulToken($typeAnalysis->getStartIndex() - 1);
+            $typeEndIndex = $typeAnalysis->getEndIndex();
+
+            if ($tokens->generatePartialCode($typeStartIndex, $typeEndIndex) !== $typeAnalysis->getName()) {
+                continue;
+            }
```

Comment:
I removed this block and unit tests for this fixer did not fail. What am I missing?

Reply:
Hard to tell, are you using [VW Extension](https://github.com/hugues-m/phpunit-vw)? When I remove that block I'm getting failed tests: https://github.com/6b7562617765726c6f73/PHP-CS-Fixer/commit/ded6e767abff695f56e4857f55b87523eee94516

Reply:
ack, it's only for 7.4